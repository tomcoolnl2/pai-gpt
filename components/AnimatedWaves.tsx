import * as React from 'react';
import * as Three from 'three';
import * as Fiber from '@react-three/fiber';
import circleImg from 'public/circle.png';

export const AnimationPoints: React.FC = () => {
	//
	const bufferRef = React.useRef<Three.BufferAttribute | null>(null);
	const imgTexture = Fiber.useLoader(Three.TextureLoader, circleImg.src);

	let t = 0;
	const frequency = 0.002;
	const amplitude = 3;

	const graph = React.useCallback(
		(x: number, z: number) => Math.sin(frequency * (x ** 2 + z ** 2 + t)) * amplitude,
		[t, frequency, amplitude],
	);

	const count = 500;
	const sep = 3;

	const bufferAttrPositions = React.useMemo(() => {
		const positions = [];
		for (let xi = 0; xi < count; xi += 1) {
			for (let zi = 0; zi < count; zi += 1) {
				let x = sep * (xi - count / 2);
				let z = sep * (zi - count / 2);
				let y = graph(x, z);
				positions.push(x, y, z);
			}
		}
		return new Float32Array(positions);
	}, [graph]);

	Fiber.useFrame(() => {
		const positions = bufferRef.current.array;
		t += 5;
		let i = 0;
		for (let xi = 0; xi < count; xi += 1) {
			for (let zi = 0; zi < count; zi += 1) {
				let x = sep * (xi - count / 2);
				let z = sep * (zi - count / 2);
				positions[i + 1] = graph(x, z);
				i += 3;
			}
		}
		bufferRef.current.needsUpdate = true;
	});

	return (
		<points>
			<bufferGeometry attach="geometry">
				<bufferAttribute
					ref={bufferRef}
					attach="attributes-position"
					array={bufferAttrPositions}
					count={bufferAttrPositions.length / 3}
					itemSize={3}
				/>
			</bufferGeometry>
			<pointsMaterial
				attach="material"
				map={imgTexture}
				color={0x00aaff}
				size={0.5}
				sizeAttenuation
				transparent={false}
				alphaTest={0.5}
				opacity={0.5}
			/>
		</points>
	);
};

export const AnimatedWaves: React.FC = () => {
	return (
		<React.Suspense fallback={null}>
			<Fiber.Canvas camera={{ position: [100, 10, 0], fov: 75 }}>
				<React.Suspense fallback={null}>
					<AnimationPoints />
				</React.Suspense>
			</Fiber.Canvas>
		</React.Suspense>
	);
};
