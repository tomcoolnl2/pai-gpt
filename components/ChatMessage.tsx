import React from 'react';
import Image from 'next/image';
import { useUser } from '@auth0/nextjs-auth0/client';
import { ConversationRole, Message } from 'context/conversation/ChatMessage';
import logo from 'public/logo.svg';

type Props = Omit<Message, 'id'>;

export const ChatMessage: React.FC<Props> = ({ role, content }) => {
	//
	const { user } = useUser();

	console.log(role);

	const avatar = React.useMemo(() => {
		const defaultSrc = 'http://www.gravatar.com/avatar';
		let src = '';
		switch (role) {
			case ConversationRole.ASSISTANT:
			case ConversationRole.SYSTEM:
				src = logo.src;
				break;
			case ConversationRole.USER:
				src = user ? user.picture : defaultSrc;
				break;
			default:
				src = defaultSrc;
		}
		return <Image src={src} width={30} height={30} alt={`${role} avatar`} className="avatar" />;
	}, [user, role]);

	return (
		<li className={`message message-${role}`}>
			{avatar}
			<p>{content}</p>
		</li>
	);
};
