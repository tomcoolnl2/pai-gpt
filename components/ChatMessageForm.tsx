import React from 'react';
import { AnswerMessage, SystemErrorMessage } from 'model';
import { useConversationContext } from 'context/conversation';

export const ChatMessageForm: React.FC<{ conversationId: null | string }> = ({ conversationId }) => {
	//
	const [question, setQuestion] = React.useState<string>('');
	const { answerStream, sendMessage } = useConversationContext();

	const disabled = React.useMemo(() => {
		if (answerStream instanceof SystemErrorMessage) {
			return true;
		}
		if (answerStream instanceof AnswerMessage) {
			return !answerStream.done;
		}
		return false;
	}, [answerStream]);

	const handleSubmit = React.useCallback(
		async (event: React.FormEvent<HTMLFormElement>) => {
			event.preventDefault();
			sendMessage(conversationId, question);
			setQuestion('');
		},
		[question, sendMessage],
	);

	const handleKeyPress = React.useCallback(
		(event: React.KeyboardEvent<HTMLTextAreaElement>) => {
			if (event.key === 'Enter' && !event.shiftKey) {
				handleSubmit(event as unknown as React.FormEvent<HTMLFormElement>);
			}
		},
		[handleSubmit],
	);

	return (
		<footer className="bg-gray-800 p-10">
			<form onSubmit={handleSubmit} noValidate>
				<fieldset className="flex gap-2" disabled={disabled}>
					<textarea
						value={question}
						onChange={(e) => setQuestion(e.target.value)}
						onKeyDown={handleKeyPress}
						placeholder="Send a message..."
						className="textarea"
					/>
					<button type="submit" className="btn">
						Send
					</button>
				</fieldset>
			</form>
		</footer>
	);
};
