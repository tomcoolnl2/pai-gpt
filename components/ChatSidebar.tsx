import Link from 'next/link';

export const ChatSidebar: React.FC = () => {
	return (
		<aside className="bg-gray-900 text-white p-10">
			<Link href="/api/auth/logout">Logout</Link>
		</aside>
	);
};
