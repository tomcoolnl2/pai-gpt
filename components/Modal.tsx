import React from 'react';
import ReactDOM from 'react-dom';
import { faClose } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

interface Props {
	open: boolean;
	dismissable?: boolean;
	handleClose: () => void;
	children: React.ReactNode;
}

export const Modal: React.FC<Props> = ({ open, dismissable = true, handleClose, children }) => {
	//
	const [documentMounted, setDocumentMounted] = React.useState<boolean>(false);

	React.useEffect(() => {
		setDocumentMounted(true);
	}, []);

	return (
		documentMounted &&
		ReactDOM.createPortal(
			<div onClick={handleClose} className={`modal-backdrop ${open ? 'open' : 'invisible'}`}>
				<div onClick={(e) => e.stopPropagation()} className={`modal${open ? ' open' : ''}`}>
					{dismissable && (
						<button onClick={handleClose} className="modal-close">
							<FontAwesomeIcon icon={faClose} />
						</button>
					)}
					{children}
				</div>
			</div>,
			document.getElementById('modal-root'),
		)
	);
};
