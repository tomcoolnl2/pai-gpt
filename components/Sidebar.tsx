import Link from 'next/link';
import React from 'react';
import { useConversationContext } from 'context/conversation';
import { Conversation } from 'model';

export const Sidebar: React.FC = () => {
	//
	const { conversations } = useConversationContext();

	return (
		<aside className="bg-gray-900 text-white">
			<Link className="btn" href="/chat">
				New Chat
			</Link>
			<ul className="flex-1 overflow-auto bg-gray-950">
				{conversations &&
					conversations.map((conversation) => (
						<li key={conversation.id} className="overflow-auto">
							<code>{conversation.id}</code>
							{conversation.title}
						</li>
					))}
			</ul>
			<Link href="/api/auth/logout">Logout</Link>
		</aside>
	);
};
