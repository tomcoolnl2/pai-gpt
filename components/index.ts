export * from './AnimatedWaves';
export * from './ChatSidebar';
export * from './ChatMessageForm';
export * from './ChatMessage';
