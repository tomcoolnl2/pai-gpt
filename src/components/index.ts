export * from './AnimatedWaves';
export * from './Spinner';
export * from './Sidebar';
export * from './SidebarItem';
export * from './Logo';
export * from './ChatMessageForm';
export * from './ChatMessage';
export * from './Modal';
export * from './DeleteConversationModal';
