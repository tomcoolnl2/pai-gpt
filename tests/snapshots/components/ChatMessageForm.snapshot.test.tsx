import { render } from '@testing-library/react';
import { ChatMessageForm } from 'components';

test('ChatMessageForm matches snapshot', () => {
	const { container } = render(<ChatMessageForm conversationId="1" />);
	expect(container).toMatchSnapshot();
});
