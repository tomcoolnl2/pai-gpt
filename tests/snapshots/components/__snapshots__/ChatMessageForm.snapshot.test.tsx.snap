// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessageForm matches snapshot 1`] = `
<div>
  <footer
    class="bg-gray-800 p-10"
  >
    <form
      novalidate=""
    >
      <fieldset
        class="flex gap-2"
      >
        <textarea
          class="textarea"
          placeholder="Send a message..."
        />
        <button
          class="btn"
          type="submit"
        >
          Send
        </button>
      </fieldset>
    </form>
  </footer>
</div>
`;
